### PR TITLE
feat(riot/lol): add option parameter to match client List()

### LINKED ISSUE #37: More options to "MatchClient"
# More options to "MatchClient"
**/lol/match/v4/matchlists/by-account/{encryptedAccountId}**

- champion
- queue
- season
- beginTime
- endTime
- beginIndex
- endIndex

It has a total of 7 optional parameters. Since the season is *DEPRECATED*, in the remaining 6 cases, the current golio library supports only beginIndex and endIndex. The rest of the parameters are also useful depending on the situation, so I think it is better to support the rest. However, if all the parameters are supported, the following method is proposed as the function argument will be greatly enlarged..

## Option Struct
```go
import (
  "github.com/KnutZuidema/golio/api"
  "github.com/KnutZuidema/golio/riot/lol"
)

func main(){
  ...
  client.Riot.LoL.Match.List(accountId, lol.MatchListOption{
    BeginIndex:1,
    EndTime: ...,
    ...
  })
}
```
## Option Function
```go
import (
  "github.com/KnutZuidema/golio/api"
  "github.com/KnutZuidema/golio/riot/lol"
  "time"
)

func main(){
  ...
  client.Riot.LoL.Match.List(accountId,
     lol.Index(0,100),
     lol.BeginTime(time.Now().Add(-time.Second*100)
  ))
}
```

## Builder
```go
import (
  "github.com/KnutZuidema/golio/api"
  "github.com/KnutZuidema/golio/riot/lol"
  "time"
)

func main(){
  ...
  client.Riot.LoL.Match.Request().
    List().
    Index(0,100).
    BeginTime(time.Now().Add(-time.Second*100).
    Do()
}
```

### PR DESCRIPTION
**Pull request by Ironpark**
resolved #37

### CODE REVIEW

File: riot/lol/match.go [hunk 1]
```diff
@@ -14,6 +14,35 @@ type MatchClient struct {
...
+// MatchListOptions
+type MatchListOptions struct {
+	// Set of champion IDs for filtering the matchlist.
+	Champion int
+	// Set of queue IDs for filtering the matchlist.
+	Queue int
+	// The begin time to use for filtering matchlist specified as epoch milliseconds.
+	BeginTime int64
+	// The end time to use for filtering matchlist specified as epoch milliseconds.
+	EndTime int64
```
Comment (KnutZuidema) [inline riot/lol/match.go:26]: For ease of use, I think it makes sense to let the user pass in a `time.Time` value here and for `BeginTime` and convert it in `buildParam`.

File: riot/lol/match.go [hunk 2]
```diff
@@ -14,6 +14,35 @@ type MatchClient struct {
 	c *internal.Client
 }
 
+// MatchListOptions
+type MatchListOptions struct {
+	// Set of champion IDs for filtering the matchlist.
+	Champion int
```
Comment (KnutZuidema) [inline riot/lol/match.go:20]: It says that it's a SET of IDs, however it's only an `int`.
```suggestion
	Champion []int
```

File: riot/lol/match.go [hunk 3]
```diff
@@ -14,6 +14,35 @@ type MatchClient struct {
...
+	// Set of queue IDs for filtering the matchlist.
+	Queue int
```
Comment (KnutZuidema) [inline riot/lol/match.go:22]: same here

File: riot/lol/match.go [hunk 4]
```diff
@@ -14,6 +14,35 @@ type MatchClient struct {
 	c *internal.Client
 }
 
+// MatchListOptions
+type MatchListOptions struct {
```
Comment (KnutZuidema) [inline riot/lol/match.go:18]: The comment is missing

File: riot/lol/match.go [hunk 5]
```diff
@@ -14,6 +16,39 @@ type MatchClient struct {
...
+			param += "&champion=" + strconv.Itoa(champion)
+		}
+	}
+	if len(mo.Queue) != 0 {
+		for _, queue := range mo.Queue {
+			param += "&queue=" + strconv.Itoa(queue)
+		}
+	}
+	if mo.BeginTime.Unix() != -62135596800 {
+		param += "&beginTime=" + strconv.FormatInt(mo.BeginTime.UnixNano()/int64(time.Millisecond), 10)
```
Comment (KnutZuidema) [inline riot/lol/match.go:44]:
```suggestion
	if !mo.BeginTime.IsZero() {
		param += "&beginTime=" + strconv.Itoa(int(mo.BeginTime.UnixNano() / 1000 / 1000))
```
I think we should use the more straightforward methods `strconv.Itoa` and `time.Time.IsZero` for this.
Same goes for the handling of `EndTime` below.

Comment (Ironpark) [reply riot/lol/match.go]: Thanks for the good advice. 
fixed!

File: riot/lol/match.go [hunk 6]
```diff
@@ -14,6 +16,39 @@ type MatchClient struct {
...
+		for _, champion := range mo.Champion {
+			param += "&champion=" + strconv.Itoa(champion)
+		}
+	}
+	if len(mo.Queue) != 0 {
+		for _, queue := range mo.Queue {
+			param += "&queue=" + strconv.Itoa(queue)
+		}
+	}
+	if mo.BeginTime.IsZero() {
```
Comment (KnutZuidema) [inline riot/lol/match.go:43]: Pretty sure you mean this:
```suggestion
	if !mo.BeginTime.IsZero() {
```
Same below

File: riot/lol/match_test.go [hunk 7]
```diff
@@ -37,7 +38,12 @@ func TestMatchClient_List(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			client := internal.NewClient(api.RegionEuropeWest, "API_KEY", tt.doer, logrus.StandardLogger())
-			got, err := (&MatchClient{c: client}).List("id", 0, 1)
+			got, err := (&MatchClient{c: client}).List("id", 0, 1, &MatchListOptions{
+				Champion:  []int{1},
+				Queue:     []int{200},
+				BeginTime: time.Time{},
+				EndTime:   time.Time{},
```
Comment (KnutZuidema) [inline riot/lol/match_test.go:45]:
```suggestion
				BeginTime: time.Now(),
				EndTime:   time.Now(),
```
We should use actual values here for coverage.

**Pull request merged**